File: 11_动态规划/子数组/回文串/132. 分割回文串 II.ts
/**
 * @param {string} s   s 的长度可以到达 2000
 * @return {number}
 * 类似于300. 最长递增子序列
 */
var minCut = function (s: string): number {
  const isPalindrome = (str: string) => str === str.split('').reverse().join('')
  const n = s.length
  const dp = Array(n).fill(n - 1)

  for (let i = 0; i < n; i++) {
    if (isPalindrome(s.slice(0, i + 1))) {
      dp[i] = 0
      continue
    }
    for (let j = 0; j < i; j++) {
      if (isPalindrome(s.slice(j + 1, i + 1))) dp[i] = Math.min(dp[i], dp[j] + 1)
    }
  }

  return dp[n - 1]
}

console.log(minCut('aab'))
// 输出：1
// 解释：只需一次分割就可将 s 分割成 ["aa","b"] 这样两个回文子串。
// 状态转移方程：dp[i] = min(dp[i], dp[j] + 1);dp[i]=min(dp[i],dp[j]+1); 如果 isPalindrome(s[j + 1..i])isPalindrome(s[j+1..i])

File: 11_动态规划/dp分类/区间dp/dfs/回文/中心扩展法求回文子串/132. 分割回文串 II.py
# 给你一个字符串 s，请你将 s 分割成一些子串，使每个子串都是回文。
# 返回符合要求的 最少分割次数

# !2472. 不重叠回文子字符串的最大数目

from 中心扩展法 import expand2


class Solution:
    def minCut(self, s: str) -> int:
        isPalindrome = expand2(s)  # !可用马拉车优化到O(n)处理+O(1)判断
        n = len(s)
        dp = [n] * (n + 1)
        dp[0] = 0
        for i in range(1, n + 1):
            for j in range(i):
                if isPalindrome[j][i - 1]:
                    dp[i] = min(dp[i], dp[j] + 1)
        return dp[-1] - 1


print(Solution().minCut("aab"))
print(Solution().minCut("ab"))
